### ISSUE TITLE
Example notebooks for parallel branch 

### ISSUE THREAD

**Issue by Ninontwik** [the-virtual-brain/tvb-multiscale#44]
Dear @dionperd,

I raised an issue last time about the use of NEST 3.1 in TVB_multiscale. 

I am using the docker image I built from the parallelNEST31 branch. Currently, I am trying to implement a spiking-neuronal network model of the thalamus into a whole-brain TVB model. Both the thalamus model in NEST, and the whole-brain model in TVB work perfectly separately with this docker image. So now I want to use tvb-multiscale to combine the two. 

I started following the workflow described in the example notebooks of the master branch, so the WilsonCowan.ipynb and the RedWongWang.ipynb but quite some errors occur because I guess there are quite some changes in functions between the paralellNEST31 and the master (for example in the default_exc_io_inh_i.py). 

There are adjusted notebooks in the parallel branch, but somehow I cannot open or download these: https://github.com/the-virtual-brain/tvb-multiscale/blob/parallelNEST31/docs/documented_example_notebooks/RedWongWang.ipynb. 
I don't know if that is a general problem or if I am doing something wrong. 

When I try adapting the old workflow, everything goes well up to building the spiking network: nest_network = nest_model_builder.build_spiking_network(). Even though I think I did everything right with respect to the output_devices I get this error:
```
TraitTypeError: Attribute can't be set to an instance of <class 'list'>
  attribute tvb_multiscale.core.spiking_models.network.SpikingNetwork.output_devices = Attr(field_type=<class 'pandas.core.series.Series'>, default=Series([], dtype: float64), required=True)
```

Maybe I did make a mistake somewhere, let me know if I should provide some of my code. But I think it would greatly help to have updated example notebooks with workflows I can follow! 

Thanks in advance and have a nice day :)

**Comment by dionperd** [comment]
Hello, sorry the late reply, but it's been a crazy week!

The error complains for entering output_devices as a list, instead of a pandas.Series.

I will have to check myself if the examples run OK in parallelNEST31, and then, if yes, as I am expecting at least the WilsonCowan one to be ok, I might have to check your code.

More, next week though!

**Comment by Ninontwik** [comment]
Hi! 

Currently I can download the notebooks in the parallel branch. However I cannot execute them, again I think because some of the functions have changed. The first error I get for example that DefaultExcIOInhIBuilder has no input argument get_defaults. Which is of course easily fixed by removing it. 

But then I again get the list error. But the example notebooks don't use pandas at all, and personally, I would also like to avoid using it haha. 

In conclusion the example notebooks do not go together with the files, paths and functions in the parallelNEST31 I think. But it would indeed be great if you could check that when you find the time! Thanks in advance!

**Comment by dionperd** [comment]
Hello,

I am not sure why you are getting these errors.

I have ran now the two [WilsonCowan](https://github.com/the-virtual-brain/tvb-multiscale/blob/parallelNEST31/examples/tvb_nest/notebooks/WilsonCowan.ipynb)([Minimal](https://github.com/the-virtual-brain/tvb-multiscale/blob/parallelNEST31/examples/tvb_nest/notebooks/WilsonCowanMinimal.ipynb)) example notebooks of tvb_nest without problems.

The [RedWongWang](https://github.com/the-virtual-brain/tvb-multiscale/blob/parallelNEST31/examples/tvb_nest/notebooks/RedWongWang.ipynb) one cannot run because the compilation of my NEST models is what is really not working yet for the latest version of NEST, and that notebook requires a  [spiking neuron model I have entered in NEST](https://github.com/the-virtual-brain/tvb-multiscale/tree/parallelNEST31/tvb_multiscale/tvb_nest/nest/modules/iaf_cond_ww_deco).

Can you tell me what are the examples that do not run in your case?

For debugging your own code, we could either have a skype call, or you would have to share it with me. I am sorry!

In all cases, I assume you are working on the [parallelNEST31 branch](https://github.com/the-virtual-brain/tvb-multiscale/tree/parallelNEST31).
It is also important that you DO run line 71 of the [Dockerfile](https://github.com/the-virtual-brain/tvb-multiscale/blob/parallelNEST31/docker/Dockerfile) in that branch to move to the [specific commit of NEST 3.1 that allows the state of nodes to change during a Prepare-Run-CleanUp cycle](https://github.com/nest/nest-simulator/commit/c545255f7252308d6e73a851ddeece4f2dffc812):
`    git fetch origin && git checkout c545255f7252308d6e73a851ddeece4f2dffc812; \
`

Can you confirm this?

**Comment by Ninontwik** [comment]
Hello,

Oh I'm sorry I had not seen this minimal WilsonCowan example. I tried the notebooks here: https://github.com/the-virtual-brain/tvb-multiscale/tree/parallelNEST31/docs/documented_example_notebooks , which do not work on the parallelNEST31 branch I think. 

If I run the WilsonCowanMinimal I get a few errors about the nest kernel, which I think is some leftover code from nest 2. So for example, 'grng_seed' and 'time' are not valid kernel parameters in nest 3.1. But this is easily fixed by adapting some code in for example devices.py. After fixing a bunch of these errors the example notebook seems to work!! Hooray! 

So thank you for your help, sorry for not finding this notebook before. 

Now I just have to figure out how to translate this example to my TVB model which uses JansenRit. That is why I also wanted to use the RedWongWang example, I feel like it is more generalized than the WilsonCowan example. But I will try to figure it out with this example and let you know if I have any more questions. 

It might be nice to delete the notebooks in this folder:  https://github.com/the-virtual-brain/tvb-multiscale/tree/parallelNEST31/docs/documented_example_notebooks since they are confusing. 

Again, thank you for your help and time. I think we can close this Issue.

**Comment by dionperd** [comment]
Hello,

I don't get any errors about the NEST kernel for any of the notebooks inside examples/tvb_nest.

So, please make sure/confirm that you are using the correct NEST version:

In all cases, I assume you are working on the parallelNEST31 branch.
It is also important that you DO run line 71 of the Dockerfile in that branch to move to the specific commit of NEST 3.1 that allows the state of nodes to change during a Prepare-Run-CleanUp cycle:
`git fetch origin && git checkout c545255f7252308d6e73a851ddeece4f2dffc812; \
`

Thanks for pointing out the outdated versions of the docs notebooks.

I have now substituted them with the newer versions from examples/tvb_nest. 

I am interested to see your progress with Jansen Rit. Feel free to contact me again about it. 

For the moment, I am closing the issue as you suggested.
